File: src/动态规划/template/背包/树上背包/P2014_CTSC1997_选课/Main2.scala
package 动态规划.template.背包.树上背包.P2014_CTSC1997_选课

import scala.collection.mutable

object Main2 {
  val N = 305
  var n: Int = 0
  var m: Int = 0
  val e: Array[mutable.ListBuffer[Int]] = Array.fill(N)(mutable.ListBuffer.empty[Int])
  val w: Array[Int] = new Array[Int](N)
  val f: Array[Array[Int]] = Array.ofDim[Int](N, N)
  val siz: Array[Int] = new Array[Int](N)

  def dfs(u: Int): Unit = {
    f(u)(1) = w(u)
    siz(u) = 1
    for (v <- e(u)) {
      dfs(v)
      siz(u) += siz(v)
      for (j <- Math.min(m + 1, siz(u)) to 1 by -1) {
        for (k <- 0 to Math.min(j - 1, siz(v))) {
          f(u)(j) = Math.max(f(u)(j), f(u)(j - k) + f(v)(k))
        }
      }
    }
  }

  def main(args: Array[String]): Unit = {
    val scanner = new java.util.Scanner(System.in)
    n = scanner.nextInt()
    m = scanner.nextInt()

    for (i <- 1 to n) {
      val k = scanner.nextInt()
      w(i) = scanner.nextInt()
      e(k) += i
    }

    dfs(0)
    println(f(0)(m + 1))
    scanner.close()
  }
}

File: src/基础算法和其他/差分/template/树上差分/Luogu_P3128/Main2.java
package 基础算法和其他.差分.template.树上差分.Luogu_P3128;

import java.util.*;

public class Main2 {
    static final int N = 50005;
    static final int M = N - 1;

    static int cnt;
    static int[] head = new int[N];
    static int[][] fa = new int[N][21];
    static int[] depth = new int[N];
    static int[] power = new int[N];
    static int ans;

    static class Edge {
        int v;
        int next;
    }

    static Edge[] edges = new Edge[M * 2];

    static {
        for (int i = 0; i < M * 2; i++) {
            edges[i] = new Edge();
        }
    }

    static void add(int u, int v) {
        cnt++;
        edges[cnt].v = v;
        edges[cnt].next = head[u];
        head[u] = cnt;
    }

    static void dfs(int son, int father) {
        fa[son][0] = father;
        depth[son] = depth[father] + 1;

        for (int i = 1; (1 << i) <= depth[son]; i++) {
            fa[son][i] = fa[fa[son][i - 1]][i - 1];
        }

        for (int i = head[son]; i != 0; i = edges[i].next) {
            int v = edges[i].v;
            if (v != father) {
                dfs(v, son);
            }
        }
    }

    static int lca(int x, int y) {
        if (depth[x] < depth[y]) {
            int temp = x;
            x = y;
            y = temp;
        }

        for (int i = 20; i >= 0; i--) {
            if (depth[fa[x][i]] >= depth[y]) {
                x = fa[x][i];
            }
        }
        if (x == y) return x;

        for (int i = 20; i >= 0; i--) {
            if (fa[x][i] != fa[y][i]) {
                x = fa[x][i];
                y = fa[y][i];
            }
        }
        return fa[x][0];
    }

    static void getAns(int son, int father) {
        for (int i = head[son]; i != 0; i = edges[i].next) {
            if (edges[i].v == father) continue;
            getAns(edges[i].v, son);
            power[son] += power[edges[i].v];
        }
        ans = Math.max(ans, power[son]);
    }


    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int k = scanner.nextInt();
        int u, v;


        for (int i = 0; i < n - 1; i++) {
            u = scanner.nextInt();
            v = scanner.nextInt();
            add(u, v);
            add(v, u);
        }

        dfs(1, 0);

        for (int i = 0; i < k; i++) {
            u = scanner.nextInt();
            v = scanner.nextInt();
            int togetherFather = lca(u, v);
            power[u]++;
            power[v]++;
            power[togetherFather]--;
            power[fa[togetherFather][0]]--;
        }

        getAns(1, 0);
        System.out.println(ans);
        scanner.close();
    }
}
